src/cpu/flexcpu/flexcpu_thread.cc

```diff
@@ -652,9 +750,76 @@ FlexCPUThread::onExecutionCompleted(shared_ptr<InflightInst> inst_ptr,
     if (inst_ptr->staticInst()->isControl()) {
         // We must have our next PC now, since this branch has just resolved
 
-        // If we are a control instruction, then we were unable to begin a
-        // fetch during decode, so we should start one now.
-        advanceInst(getNextPC());
+        if (_cpuPtr->hasBranchPredictor()) {
+            // If a branch predictor has been set, then we need to check if
+            // a prediction was made, squash instructions if incorrectly
+            // predicted, and notify the predictor accordingly.
+
+            auto it = find(inflightInsts.begin(), inflightInsts.end(),
+                           inst_ptr);
+
+            if (it == inflightInsts.end()) {
+                // Should be unreachable...
+                panic("Did a completion callback on an instruction not in "
+                      "our buffer.");
+            }
+
+            shared_ptr<InflightInst> following_inst =
+                (++it != inflightInsts.end()) ? *it : nullptr;
+
+            if (following_inst || advanceInstNeedsRetry) {
+                // This condition is true iff a branch prediction was made
+                // prior to this point in simulation for this instruction.
+
+                freeBranchPredDepth();
+
+                const TheISA::PCState calculatedPC = inst_ptr->pcState();
+                TheISA::PCState correctPC = calculatedPC;
+                inst_ptr->staticInst()->advancePC(correctPC);
+                const TheISA::PCState predictedPC = following_inst ?
+                    following_inst->pcState() : advanceInstRetryPC;
+
+                if (correctPC.pc() == predictedPC.pc()
+                 && correctPC.upc() == predictedPC.upc()) {
+                    // It was a correct prediction
+                    DPRINTF(FlexCPUBranchPred,
+                            "Branch predicted correctly (seq %d)\n",
+                            inst_ptr->seqNum());
+                    // Calculate stat correct predictions?
+                    // Otherwise do nothing because we guessed correctly.
+                    // Just remember to update the branch predictor at commit.
```

Comment:
Why not update the branch predictor now? This could be a reason we have so many incorrect branches...

Reply:
We don't know that this branch will even be committed.

We could have like ld r1, 100 -> br r1,16 (mispredict?) ->ld r2, 200 -> br r2, 32 (correct?), where some previous instruction allows r2 to be read before r1. the second branch completes execution first, but we don't know if its "correct" prediction is actually even part of code which should be executed.